### PR TITLE
chore(cairo_native): cache both casm and sierra in the same global cache

### PR DESCRIPTION
**Pull request by avi-starkware**

### CODE REVIEW

File: crates/papyrus_state_reader/src/papyrus_state.rs [hunk 1]
```diff
@@ -134,16 +135,22 @@ impl StateReader for PapyrusReader {
...
-                self.contract_class_manager
-                    .set_casm(class_hash, versioned_contract_class_from_db.clone());
-                Ok(RunnableCompiledClass::from(versioned_contract_class_from_db))
+                self.contract_class_manager.set_casm(class_hash, cachable_casm);
+                Ok(RunnableCompiledClass::from(versioned_compiled_class_from_db))
+            }
+            Some(CachedCasm::WithoutSierra(versioned_compiled_class)) => {
+                Ok(RunnableCompiledClass::from(versioned_compiled_class))
+            }
+            Some(CachedCasm::WithSierra) => {
```
[the remark this answers is not in the document]
Comment (avi-starkware) [reply crates/papyrus_state_reader/src/papyrus_state.rs]: Done.

**Review by meship-starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: all files reviewed, 2 unresolved discussions (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

___
*[`crates/blockifier/src/state/global_cache.rs` line 5 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEikcr-ChhdGGGTY7df:-OEikcr-ChhdGGGTY7dg:byp18fd) ([raw file](https://github.com/starkware-libs/sequencer/blob/16c59879d4d9b2516311c1035a2d6125343c932d/crates/blockifier/src/state/global_cache.rs#L5)):*
> ```XML
> use cached::{Cached, SizedCache};
> use starknet_api::core::ClassHash;
> #[cfg(feature = "cairo_native")]
> ```

Remove



_Code quote:_
```XML
#[cfg(feature = "cairo_native")]
```
___
*[`crates/blockifier/src/state/global_cache.rs` line 73 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEikwMR0Mddp12r5FIe:-OEikwMS91k04lirLL8r:bl6ew1k) ([raw file](https://github.com/starkware-libs/sequencer/blob/16c59879d4d9b2516311c1035a2d6125343c932d/crates/blockifier/src/state/global_cache.rs#L73)):*
> ```XML
>     pub fn set_casm(&self, class_hash: ClassHash, compiled_class: CachedCasm) {
>         self.casm_cache.set(class_hash, compiled_class);
>     }
> ```

I am not sure about it because it is only one class without the Sierra



_Suggestion:_
```XML
    pub fn set_casm(&self, class_hash: ClassHash, compiled_classes: CachedCasm) {
        self.casm_cache.set(class_hash, compiled_class);
    }
```

**Comment by avi-starkware** [comment]
*[`crates/blockifier/src/state/global_cache.rs` line 5 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEikcr-ChhdGGGTY7df:-OEio1va3cUf-UdDgKGj:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/16c59879d4d9b2516311c1035a2d6125343c932d/crates/blockifier/src/state/global_cache.rs#L5)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Remove
</blockquote></details>

Done.

**Review by meship-starkware — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: all files reviewed, 1 unresolved discussion (waiting on @avi-starkware, @avivg-starkware, and @noaov1)

**Comment by avi-starkware** [comment]
*[`crates/blockifier/src/state/global_cache.rs` line 73 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEikwMR0Mddp12r5FIe:-OEiowqN5qhiOZlHA37h:b-phcqmy) ([raw file](https://github.com/starkware-libs/sequencer/blob/16c59879d4d9b2516311c1035a2d6125343c932d/crates/blockifier/src/state/global_cache.rs#L73)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

I am not sure about it because it is only one class without the Sierra
</blockquote></details>

I would rather keep the name `compiled_class`.
There is only one compiled class. The cached entry just might include the sierra contract class in addition. (Similarly, the cairo_native struct includes the casm but we don't call it compiled_classes plural.)

**Review by meship-starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware and @noaov1)

**Review by noaov1 — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: all files reviewed, 2 unresolved discussions (waiting on @avi-starkware and @avivg-starkware)

___
*[`crates/papyrus_state_reader/src/papyrus_state.rs` line 140 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEj0XcW7dNtgJBx5U1a:-OEj0XcX2FFsd3vV2ONl:b-sybkjw) ([raw file](https://github.com/starkware-libs/sequencer/blob/e166e2cc294de9080264762680ee5c3296b4ebe1/crates/papyrus_state_reader/src/papyrus_state.rs#L140)):*
> ```XML
>         match cached_casm {
>             None => {
>                 let compiled_class_from_db = self.get_compiled_class_inner(class_hash)?;
> ```

Are you planning to change to change `get_compiled_class_inner` to return the sierra as well?



_Code quote:_
```XML
                let compiled_class_from_db = self.get_compiled_class_inner(class_hash)?;
```
___
*[`crates/papyrus_state_reader/src/papyrus_state.rs` line 149 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEj10UIDNy0eGWDBFSt:-OEj10UIDNy0eGWDBFSu:b-ac1xe2) ([raw file](https://github.com/starkware-libs/sequencer/blob/e166e2cc294de9080264762680ee5c3296b4ebe1/crates/papyrus_state_reader/src/papyrus_state.rs#L149)):*
> ```XML
>             Some(CachedCasm::WithoutSierra(casm)) => Ok(casm),
>             Some(CachedCasm::WithSierra(_, _)) => {
>                 todo!("Add this flow when Sierra to Native compilation is added to PapyrusReader.")
> ```

WDYM? Why not simply return the casm?



_Code quote:_
```XML
            Some(CachedCasm::WithSierra(_, _)) => {
                todo!("Add this flow when Sierra to Native compilation is added to PapyrusReader.")
```

**Review by avi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: all files reviewed, 2 unresolved discussions (waiting on @avivg-starkware and @noaov1)

___
*[`crates/papyrus_state_reader/src/papyrus_state.rs` line 140 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEj0XcW7dNtgJBx5U1a:-OEj6yXa9UyhUTwU2Bbc:bhsswcb) ([raw file](https://github.com/starkware-libs/sequencer/blob/e166e2cc294de9080264762680ee5c3296b4ebe1/crates/papyrus_state_reader/src/papyrus_state.rs#L140)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

Are you planning to change to change `get_compiled_class_inner` to return the sierra as well?
</blockquote></details>

This would be part of meshi's PR.
Currently she returns the sierra as Option in her PR.

I just implemented this method to behave as it did before we added the contract class manager.
___
*[`crates/papyrus_state_reader/src/papyrus_state.rs` line 149 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2863#-OEj10UIDNy0eGWDBFSt:-OEj6Ma8-PFaRbmIvmDh:bclhtn8) ([raw file](https://github.com/starkware-libs/sequencer/blob/e166e2cc294de9080264762680ee5c3296b4ebe1/crates/papyrus_state_reader/src/papyrus_state.rs#L149)):*
<details><summary><i>Previously, noaov1 (Noa Oved) wrote…</i></summary><blockquote>

WDYM? Why not simply return the casm?
</blockquote></details>

It should be handled according to the flows meshi defines. This enum option should not be used before meshi's PR is in, so I don't think this case should be defined in this PR.

**Review by noaov1 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2863)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @avivg-starkware)

**Pull request merged**